### ISSUE TITLE
[Question] How to get a event in JS if some JSON property changes?

### ISSUE THREAD

**Issue by jwillmer** [Starcounter/Starcounter#3145]
Is there a way to register for some property change event in JS?

**Comment by jwillmer** [comment]
Have it, I use this method to observe changes in my viewmodel:

```js
<script model="{{model}}">
    (function () {
        var scriptElement = document.currentScript;
        var model = scriptElement.model;
    
        Object.observe(model, function (changes) {
            changes.forEach(function (change) {
                console.log(change.type, change.name, change.oldValue);
            });
    
        });
    
    }());
</script>
```

**Comment by Starcounter-Jack** [comment]
If this is Raphael Weinsteins Ecmascript ```Object.observe()```, it has been rejected by Ecma and will be removed, right @warpech?

**Comment by Starcounter-Jack** [comment]
@jwillmer 
We made an alternative to Object.observe() that works on all browser. 

@warpech How do we add a callback function to a viewmodel without adding a new listener? Should we add a function for this. It would be to expensive to have more than one. It would be simple for json patch to call multiple callbacks for an observed object and the viewmodel already have an attached observer using github.com/starcounter-jack/jsonpatch

**Comment by warpech** [comment]
Yes, `Object.observe()` is deprecated. The polyfill for it is only temporarily distributed with Starcounter for the internal use of `dom-bind-notifier`.

I recommend to go the Web Components way, which is to use a Custom Element that is able to observe changes in the external array. We need to add some example of that to [KitchenSink](https://github.com/StarcounterSamples/KitchenSink/blob/master/src/KitchenSink/wwwroot/KitchenSink/ChartPage.html).

> @warpech How do we add a callback function to a viewmodel without adding a new listener? Should we add a function for this. It would be to expensive to have more than one. It would be simple for json patch to call multiple callbacks for an observed object and the viewmodel already have an attached observer using http://www.github.com/starcounter-jack/json-patch

This seems also like a nice idea. There is only one problem - when the template that has registered the callback is detached from DOM, it needs to be implicitly removed. @tomalec WDYT?

**Comment by jwillmer** [comment]
Implement it to be available in web components is a good thing but please provide a way to use it with plain javascript as well, not all make sense to be implemented with web components ;-)

**Comment by warpech** [comment]
Good point. But maybe we can go with what is there already. How about this?

```html
<template>
    <template is="dom-bind">
        <!-- your stuff here-->

        <dom-bind-notifier path="model" observed-object="{{model}}" 
         check="{{checkForNewValue(model.ObservedProperty)}}" deep></dom-bind-notifier>
    </template>
    <script>
        (function init(template) {
            var old;
            template.checkForNewValue = function (newVal) {
                if (newVal != old) {
                    console.log("I am new!", newVal);
                    old = newVal;
                }
                return old;
            };
        })(document.currentScript.previousElementSibling);
    </script>
</template>
```

**Comment by jwillmer** [comment]
I like it - is it a proposal or is it already usable?

**Comment by warpech** [comment]
I just tested that it works, so feel free to use it and please let me know if you find it suitable! I will then add it to the KitchenSink.

**Comment by tomalec** [comment]
I really like the computed property way @warpech proposed. We may consider adding some sugar for it in dom-bind-notifier later - after we will finish polymer vs aurelia template binding competition and/or workaround Polymer API issues.

**Comment by warpech** [comment]
@tomalec would it be possible to get a callback as proposed by @Starcounter-Jack?

**Comment by jwillmer** [comment]
:+1: for an API

**Comment by tomalec** [comment]
If Polymer will work as expected, and if I get you correctly, we should be able to reuse puppet's (fast-json-patch) listener, so every part of DOM/JS may attach its callback for a change, but I thought about also adding listeners, but not for an object, but regular DOM event (triggered by single `<puppet-*component*>`

**Comment by miyconst** [comment]
There is another option:

```js
var script = document._currentString || document.currentScript;
var template = script.previousElementSibling;

template.messageChanged = function(newMessageValue) {
    console.log("Message was changed", newMessageValue);
};

template.observers = ["messageChanged(model.Message)"];
```

**Comment by miyconst** [comment]
@jwillmer, @warpech, @tomalec, can we close this issue as answered?

**Comment by jwillmer** [comment]
I would like to see this solutions in the wiki before closing because it is very helpful.

**Comment by warpech** [comment]
I think that @miyconst's solution is the best one. 

Thank you @jwillmer for caring about improving our docs. This is exactly what we need. I will set up a task to write that. I think that this needs to be a subpage on http://starcounter.io
